### PR TITLE
[WIP] Document the repository signatures protocol change

### PR DESCRIPTION
**Pull request by joelverhagen**
Address https://github.com/NuGet/Engineering/issues/1188.

Preview link:
https://review.docs.microsoft.com/en-us/nuget/api/repository-signatures-resource?branch=jver-reposig

/cc @zhili1208, @PatoBeltran

### CODE REVIEW

File: docs/api/_data/repository-signatures-index.json [hunk 1]
```diff
@@ -0,0 +1,15 @@
+{
+  "allRepositorySigned": false,
+  "signingCertificates": [
+    {
+      "fingerprints": {
+        "2.16.840.1.101.3.4.2.1": "79851da778a4780d2720588901a41b70ad9e584f847d09359949862f138e92ac"
```
Comment (joelverhagen) [inline docs/api/_data/repository-signatures-index.json]: This test data will be updated once we know the real certificate. This is not critical but makes a better example.

**Comment by saldana** [comment]
### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/TOC.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/jver-reposig/docs/TOC.md) | :white_check_mark:Succeeded | |
[docs/api/_data/repository-signatures-index.json](https://github.com/NuGet/docs.microsoft.com-nuget/blob/jver-reposig/docs/api/_data/repository-signatures-index.json) | :white_check_mark:Succeeded | |
[docs/api/overview.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/jver-reposig/docs/api/overview.md) | :white_check_mark:Succeeded | |
[docs/api/report-abuse-resource.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/jver-reposig/docs/api/report-abuse-resource.md) | :white_check_mark:Succeeded | |
[docs/api/repository-signatures-resource.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/jver-reposig/docs/api/repository-signatures-resource.md) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2018%5C3%5C3%5C19a7edc0-f91d-0fd5-6e5f-4d131db05c11%5CPullRequest%5C201803030018091300-723%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=nIFbnMMhZpxySMbCF5e%2BhC3oMIyqnts9WFrhdz4Rhnw%3D&st=2018-03-03T00%3A14%3A01Z&se=2018-04-03T00%3A19%3A01Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

File: docs/api/repository-signatures-resource.md [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+keywords: NuGet API repository signatures, nuget.org signing certificates, nuget.org package signing
+ms.reviewer:
+- karann
+- unniravindranathan
+---
+
+# Repository signatures
+
+If a package source supports adding repository signatures to published packages, it is possible for a client to
+determine the signing certificates that are be used by the package source. This resource allows clients to detect
```
Comment (dtivel) [inline docs/api/repository-signatures-resource.md]: "...that are ~be~ used by..."

File: docs/api/repository-signatures-resource.md [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+## Repository signatures index
+
+The repository signatures index contains two pieces of information:
+
+1. Whether or not all packages found on the source are repository signed.
+1. The list of certificates used by the package source to sign packages.
+
+In most cases, the list of certificates will only ever be appended to. New certificates would be added to the list when
+the previous signing certificate has expired and the package source needs to start using a new signing certificate. If
+a certificate is removed from the list, that means that all packages signed with the removed signing certificate should
```
Comment (dtivel) [inline docs/api/repository-signatures-resource.md]: ..."that means that all **package signatures created** with the removed signing certificate should **no** [~not~] longer be considered valid by the client."

In this case the package signature is invalid not necessarily the package.  A client policy may allow installing the package as unsigned.

File: docs/api/repository-signatures-resource.md [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+Name                | Type             | Required
+------------------- | ---------------- | --------
+allRepositorySigned | boolean          | yes
+signingCertificates | array of objects | yes
+
+The `allRepositorySigned` boolean is set to false if the package source has some packages that have no repository
+signature. If the boolean is set to true, all packages available on the source must have a repository
+signature produced by one of the signing certificates mentioned in `signingCertificates`.
+
+There may be zero or more signing certificates in the `signingCertificates` array. Each element in this array is a JSON
```
Comment (dtivel) [inline docs/api/repository-signatures-resource.md]: There must be at least one signing certificate if `allRepositorySigned` is true?

File: docs/api/repository-signatures-resource.md [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+fingerprints | object | yes      |
+subject      | string | yes      | The subject distinguished name from the certificate
+issuer       | string | yes      | The distinguished name of the certificate's issuer
+notBefore    | string | yes      | The starting timestamp of the certificate's validity period
+notAfter     | string | yes      | The ending timestamp of the certificate's validity period
+
+Note that the `contentUrl` is required to be served over HTTPS. This URL has no specific URL pattern and should be
+dynamically discovered using this repository signatures index document. 
+
+All properties in this object (aside from `contentUrl`) should be derivable from the certificate found at `contentUrl`.
```
Comment (dtivel) [inline docs/api/repository-signatures-resource.md]: should or must?

File: docs/api/repository-signatures-resource.md [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+The `fingerprints` object has the following properties:
+
+Name                   | Type   | Required | Notes
+---------------------- | ------ | -------- | -----
+2.16.840.1.101.3.4.2.1 | string | yes      | The SHA-256 fingerprint
+
+The key name `2.16.840.1.101.3.4.2.1` is the OID of the SHA-256 hash algorithm.
+
+All hash values are lowercase, hex-encoded string representations of the hash digest.
+
```
Comment (dtivel) [inline docs/api/repository-signatures-resource.md]: are or must be?

File: docs/api/repository-signatures-resource.md [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+## HTTP methods
+
+All URLs found in the repository signatures resource support only the HTTP methods `GET` and `HEAD`.
+
+## Repository signatures index
+
+The repository signatures index contains two pieces of information:
+
+1. Whether or not all packages found on the source are repository signed.
```
Comment (dtivel) [inline docs/api/repository-signatures-resource.md]: "...are repository signed **by this repository**"?

File: docs/api/repository-signatures-resource.md [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+The `allRepositorySigned` boolean is set to false if the package source has some packages that have no repository
+signature. If the boolean is set to true, all packages available on the source must have a repository
+signature produced by one of the signing certificates mentioned in `signingCertificates`.
+
+There may be zero or more signing certificates in the `signingCertificates` array. Each element in this array is a JSON
+object with the following properties.
+
+Name         | Type   | Required | Notes
+------------ | ------ | -------- | -----
+contentUrl   | string | yes      | Absolute URL to the DER encoded public certificate
```
Comment (dtivel) [inline docs/api/repository-signatures-resource.md]: DER-encoded

File: docs/api/repository-signatures-resource.md (hunk 5 above)
Comment (joelverhagen) [reply docs/api/repository-signatures-resource.md]: `must`. Thanks.

File: docs/api/repository-signatures-resource.md (hunk 6 above)
Comment (joelverhagen) [reply docs/api/repository-signatures-resource.md]: `must be`

File: docs/api/repository-signatures-resource.md (hunk 4 above)
Comment (joelverhagen) [reply docs/api/repository-signatures-resource.md]: Spoke offline. This is a possible state.

File: docs/api/repository-signatures-resource.md (hunk 8 above)
Comment (joelverhagen) [reply docs/api/repository-signatures-resource.md]: Fixed.

File: docs/api/repository-signatures-resource.md (hunk 7 above)
Comment (joelverhagen) [reply docs/api/repository-signatures-resource.md]: Fixed.

File: docs/api/repository-signatures-resource.md (hunk 6 above)
Comment (joelverhagen) [reply docs/api/repository-signatures-resource.md]: Fixed.

File: docs/api/repository-signatures-resource.md (hunk 5 above)
Comment (joelverhagen) [reply docs/api/repository-signatures-resource.md]: Fixed.

File: docs/api/repository-signatures-resource.md (hunk 4 above)
Comment (joelverhagen) [reply docs/api/repository-signatures-resource.md]: Fixed.

File: docs/api/repository-signatures-resource.md (hunk 3 above)
Comment (joelverhagen) [reply docs/api/repository-signatures-resource.md]: Fixed.

File: docs/api/repository-signatures-resource.md (hunk 2 above)
Comment (joelverhagen) [reply docs/api/repository-signatures-resource.md]: Fixed.

**Comment by saldana** [comment]
### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/TOC.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/jver-reposig/docs/TOC.md) | :white_check_mark:Succeeded | |
[docs/api/_data/repository-signatures-index.json](https://github.com/NuGet/docs.microsoft.com-nuget/blob/jver-reposig/docs/api/_data/repository-signatures-index.json) | :white_check_mark:Succeeded | |
[docs/api/overview.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/jver-reposig/docs/api/overview.md) | :white_check_mark:Succeeded | |
[docs/api/report-abuse-resource.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/jver-reposig/docs/api/report-abuse-resource.md) | :white_check_mark:Succeeded | |
[docs/api/repository-signatures-resource.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/jver-reposig/docs/api/repository-signatures-resource.md) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2018%5C3%5C5%5C19a7edc0-f91d-0fd5-6e5f-4d131db05c11%5CPullRequest%5C201803051817592676-723%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=AjkQwUHg%2ByJt1wXMBCZ13F7Yvq1pXdjk9BlYHI4EURw%3D&st=2018-03-05T18%3A14%3A04Z&se=2018-04-05T18%3A19%3A04Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

**Comment by joelverhagen** [comment]
We are waiting on two things before merging this:

- [ ] Put the real signing certificate details in `docs/api/_data/repository-signatures-index.json`
- [ ] Release of (any) repository signed packages to NuGet.org.

File: docs/api/_data/repository-signatures-index.json (hunk 1 above)
Comment (joelverhagen) [reply docs/api/_data/repository-signatures-index.json]: This is done now.

**Comment by karann-msft** [comment]
### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/TOC.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/jver-reposig/docs/TOC.md) | :white_check_mark:Succeeded | |
[docs/api/_data/repository-signatures-index.json](https://github.com/NuGet/docs.microsoft.com-nuget/blob/jver-reposig/docs/api/_data/repository-signatures-index.json) | :white_check_mark:Succeeded | |
[docs/api/overview.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/jver-reposig/docs/api/overview.md) | :white_check_mark:Succeeded | |
[docs/api/report-abuse-resource.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/jver-reposig/docs/api/report-abuse-resource.md) | :white_check_mark:Succeeded | |
[docs/api/repository-signatures-resource.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/jver-reposig/docs/api/repository-signatures-resource.md) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2018%5C4%5C20%5C19a7edc0-f91d-0fd5-6e5f-4d131db05c11%5CPullRequest%5C201804201808018522-723%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=hc1x%2Bkc%2FlG%2FAi2opcCEYikljZMRkFJZhRwKF2mhonvE%3D&st=2018-04-20T18%3A05%3A50Z&se=2018-05-21T18%3A10%3A50Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

**Comment by rido-min** [comment]
@joelverhagen  We should include the NuGet.org certificate details: CN, Thumbprint, Expiration Date etc..

**Comment by karann-msft** [comment]
### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/TOC.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/jver-reposig/docs/TOC.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/nuget/api/report-abuse-resource?branch=pr-en-us-723) |
[docs/api/_data/repository-signatures-index.json](https://github.com/NuGet/docs.microsoft.com-nuget/blob/jver-reposig/docs/api/_data/repository-signatures-index.json) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/nuget/api/repository-signatures-resource?branch=pr-en-us-723) |
[docs/api/overview.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/jver-reposig/docs/api/overview.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/nuget/api/overview?branch=pr-en-us-723) |
[docs/api/report-abuse-resource.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/jver-reposig/docs/api/report-abuse-resource.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/nuget/api/report-abuse-resource?branch=pr-en-us-723) |
[docs/api/repository-signatures-resource.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/jver-reposig/docs/api/repository-signatures-resource.md) | :white_check_mark:Succeeded | [View](https://review.docs.microsoft.com/en-us/nuget/api/repository-signatures-resource?branch=pr-en-us-723) |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2018%5C8%5C8%5C19a7edc0-f91d-0fd5-6e5f-4d131db05c11%5CPullRequest%5C201808082100417276-723%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=oJjiSOOtDLCd9%2FETeZHXWSosAg6SqDLLXxIiOX32LtI%3D&st=2018-08-08T20%3A56%3A46Z&se=2018-09-08T21%3A01%3A46Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

**Pull request merged**